samples/net/http_server/src/main.c

```diff
@@ -167,6 +176,65 @@ static int http_response_soft_404(struct http_server_ctx *ctx)
 			     HTML_FOOTER);
 }
 
+static int http_response_auth(struct http_server_ctx *ctx)
+{
+	return http_response(ctx, HTTP_STATUS_200_OK, HTML_HEADER
+			     "<h2><center>user: "HTTP_AUTH_USERNAME
+			     ", password: "HTTP_AUTH_PASSWORD
+			     "</center></h2>" HTML_FOOTER);
+}
+
+int http_response_401(struct http_server_ctx *ctx, s32_t timeout)
+{
+	/* Wait 2 secs for the reply with proper credentials */
+	return http_response_wait(ctx, HTTP_401_STATUS_US, NULL, timeout);
+}
+
+static int http_basic_auth(struct http_server_ctx *ctx)
+{
+	const char auth_str[] = "Authorization: Basic ";
+	char *ptr;
+
+	ptr = strstr(ctx->req.field_values[0].key, auth_str);
+	if (ptr) {
+		char output[sizeof(HTTP_AUTH_USERNAME) +
+			    sizeof(":") +
+			    sizeof(HTTP_AUTH_PASSWORD)];
+		size_t olen, ilen, alen;;
+		char *end, *colon;
+		int ret;
+
+		memset(output, 0, sizeof(output));
+
+		end = strstr(ptr, HTTP_CRLF);
+		if (!end) {
+			return http_response_401(ctx, K_NO_WAIT);
+		}
+
+		alen = sizeof(auth_str) - 1;
+		ilen = end - (ptr + alen);
+
+		ret = mbedtls_base64_decode(output, sizeof(output) - 1,
+					    &olen, ptr + alen, ilen);
+		if (ret) {
+			return http_response_401(ctx, K_NO_WAIT);
+		}
+
+		colon = strstr(output, ":");
```

Comment:
Better to use `memchr()` here. It's not only faster, you can pass the length of `output`.

Also, null-check the output of this function call. Checking for `colon > output` below [isn't sufficient](http://c-faq.com/null/varieties.html).